### PR TITLE
First version of For Loop implementation

### PR DESCRIPTION
**Pull request by Just-Kiel**
## Description
If you connect a list to an attribute, you can iterate over the list as a for loop.
There is a expand loop button and a collapse one.
For the moment it is not possible to compute all at once but you can see every step.

### CODE REVIEW

File: meshroom/ui/graph.py [hunk 1]
```diff
@@ -782,13 +837,30 @@ def removeEdge(self, edge):
...
     @Slot(Attribute)
     def resetAttribute(self, attribute):
         """ Reset 'attribute' to its default value """
+        # if the attribute is a ListAttribute, remove all edges
+        if isinstance(attribute, ListAttribute):
+            for edge in self._graph.edges:
+                # if the edge is connected to one of the ListAttribute's elements, remove it
+                if edge.src in attribute.value:
+                    self.removeEdge(edge)
         self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))
```
Comment (fabiencastan) [inline meshroom/ui/graph.py:864]: undo/redo

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 2]
```diff
@@ -392,6 +392,11 @@ Item {
                         tooltipText: "Iterations"
                         visible: edgeMenu.currentEdge && edgeMenu.forLoop
 
+                        onVisibleChanged: {
+                            if (visible)
+                                value = listAttr.value.indexOf(edgeMenu.currentEdge.src) + 1
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/GraphEditor.qml:397]: assignations replace/break expressions

Comment (fabiencastan) [reply meshroom/ui/qml/GraphEditor/GraphEditor.qml:397]: value: (listAttr && listAttr.value) ? listAttr.value.indexOf(edgeMenu.currentEdge.src) + 1 : 0

File: meshroom/ui/graph.py [hunk 3]
```diff
@@ -857,11 +857,14 @@ def resetAttribute(self, attribute):
...
-                    self.removeEdge(edge)
-        self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))
+            with self.groupedGraphModification("Reset List Attribute '{}'".format(attribute.name)):
+                for edge in self._graph.edges:
+                    # if the edge is connected to one of the ListAttribute's elements, remove it
+                    if edge.src in attribute.value:
+                        self.removeEdge(edge)
+                self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))
+        else:
+            self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))
```
Comment (fabiencastan) [inline meshroom/ui/graph.py]:
```
    with self.groupedGraphModification("Reset Attribute '{}'".format(attribute.name)):
        if isinstance(attribute, ListAttribute):
            for edge in self._graph.edges:
                # if the edge is connected to one of the ListAttribute's elements, remove it	                for edge in self._graph.edges:
                if edge.src in attribute.value:	                    # if the edge is connected to one of the ListAttribute's elements, remove it
                    self.removeEdge(edge)	                    if edge.src in attribute.value:
        self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))	                        self.removeEdge(edge)
                self.push(commands.SetAttributeCommand(self._graph, attribute, attribute.defaultValue()))
```

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml (hunk 2 above)
Comment (Just-Kiel) [reply meshroom/ui/qml/GraphEditor/GraphEditor.qml]: I tried to change few things

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**